lib/semian/dual_circuit_breaker.rb

```diff
@@ -0,0 +1,163 @@
+# frozen_string_literal: true
+
+module Semian
+  # DualCircuitBreaker wraps both legacy and adaptive circuit breakers,
+  # allowing runtime switching between them via a callable that determines which to use.
+  class DualCircuitBreaker
+    attr_reader :name, :legacy_circuit_breaker, :adaptive_circuit_breaker
+
+    # use_adaptive should be a callable (Proc/lambda) that returns true/false
+    # to determine which circuit breaker to use. If it returns true, use adaptive.
+    def initialize(name:, legacy_circuit_breaker:, adaptive_circuit_breaker:, use_adaptive:)
+      @name = name
+      @legacy_circuit_breaker = legacy_circuit_breaker
+      @adaptive_circuit_breaker = adaptive_circuit_breaker
+      @use_adaptive = use_adaptive || ->() { false } # Default to legacy
+    end
+
+    # Main acquire method - delegates to the active circuit breaker
+    def acquire(resource = nil, &block)
+      active_circuit_breaker.acquire(resource, &block)
+    end
+
+    # State query methods - delegate to active circuit breaker
+    def open?
+      active_circuit_breaker.open?
+    end
+
+    def closed?
+      active_circuit_breaker.closed?
+    end
+
+    def half_open?
+      active_circuit_breaker.half_open?
+    end
+
+    def request_allowed?
+      active_circuit_breaker.request_allowed?
+    end
+
+    # Mark methods - record on BOTH circuit breakers for data consistency
+    def mark_failed(error)
+      @legacy_circuit_breaker&.mark_failed(error)
+      @adaptive_circuit_breaker&.mark_failed(error)
+    end
+
+    def mark_success
+      @legacy_circuit_breaker&.mark_success
+      @adaptive_circuit_breaker&.mark_success
+    end
+
+    def mark_rejected
+      @legacy_circuit_breaker&.mark_rejected if @legacy_circuit_breaker&.respond_to?(:mark_rejected)
+      @adaptive_circuit_breaker&.mark_rejected
+    end
+
+    # Stop both circuit breakers
+    def stop
+      @legacy_circuit_breaker&.stop if @legacy_circuit_breaker&.respond_to?(:stop)
+      @adaptive_circuit_breaker&.stop
+    end
+
+    # Reset both circuit breakers
+    def reset
+      @legacy_circuit_breaker&.reset
+      @adaptive_circuit_breaker&.reset
+    end
+
+    # Destroy both circuit breakers
+    def destroy
+      @legacy_circuit_breaker&.destroy
+      @adaptive_circuit_breaker&.destroy
+    end
+
+    # Check if either circuit breaker is in use
+    def in_use?
+      (@legacy_circuit_breaker&.in_use? || false) ||
+        (@adaptive_circuit_breaker&.in_use? || false)
+    end
+
+    # Get the last error from the active circuit breaker
+    def last_error
+      active_circuit_breaker.last_error
+    end
+
+    # Get metrics from both circuit breakers for comparison
+    def metrics
+      {
+        active: use_adaptive? ? :adaptive : :legacy,
+        legacy: legacy_metrics,
+        adaptive: adaptive_metrics,
+      }
+    end
+
+    # Get state from the active circuit breaker
+    def state
```

Comment:
do we need the dual circuit breaker to implement this, and the other functions below?